pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,228 @@
+                              handler_(handler),
+                              controller_read_fd_(controller_read_fd) {
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(kInfo, "Watching the root process to exit if it does.");
+
+  ssize_t status;
+  char buf[1];
+  do {
+    status = read(controller_read_fd_, buf, 1);
+  } while (status == -1 && (errno == EINTR ||
+                            errno == EAGAIN));
+  if (status == -1) {
+    handler_->Message(
+        kWarning, "Controller got error %d reading from pipe, shutting down",
+        static_cast<int>(status));
+  } else if (status == 0 /* EOF */) {
+    handler_->Message(kInfo, "Root process exited; controller shutting down.");
+  } else if (status == 1 /* read a byte */) {
+    handler_->Message(
+        kInfo, "Root process is starting a new controller; shutting down.");
+  } else {
+    LOG(FATAL) << "Status of " << status << " doesn't make sense";
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+void ControllerManager::RunController() {
+  handler_->Message(kInfo, "Controller running with PID %d", getpid());
+
+  // Would set up gRPC server here and pass control to its event loop.  Instead
+  // just hang out sleeping until the ProcessDeathWatcherThread decides it's
+  // time to exit.
+  while (true) {
+    sleep(60);  // 1m
+  }
+}
+
+void ControllerManager::Daemonize() {
+  // Make a new session (process group).
+  if (setsid() < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to setsid().");
+  }
+
+  // We need to fork again to make sure there is no session group leader.
+  pid_t pid = fork();
+  CHECK(pid != -1) << "Couldn't fork to daemonize.";
+  if (pid != 0) {
+    exit(EXIT_SUCCESS);
+  }
+
+  // If we keep the current directory we might keep them from being able to
+  // unmount their filesystem.
+  if (chdir("/") < 0) {
+    handler_->Message(kWarning, "Daemonize: Failed to chdir(/).");
+  }
+
+  // If we disconnect file descriptors then logging will break, so don't.
+}
+
+void ControllerManager::ForkOffControllerProcess() {
+  handler_->Message(kInfo, "Forking controller process off of %d", getpid());
+
+  // Whenever we fork off a controller we save the fd for a pipe to it.  Then if
+  // we fork off another controller we can write a byte to the pipe to tell the
+  // old controller to clean up and exit.
+  static int controller_write_fd = -1;
+  if (controller_write_fd != -1) {
+    // We already forked off a controller earlier.  Tell it to quit by writing a
+    // byte.  If there's no one still with the pipe open we'll get SIGPIPE and
+    // die horribly, but the babysitter-restarting-controller structure makes
+    // that very unlikely.
```

Comment:
I agree it's not desirable.  But  as long as the babysitter hasn't died it won't happen.  The babysitter dies surprisingly if it can't `fork`, can't use `waitpid`, or if the controller calls `exit` without being asked to quit, and those shouldn't happen.